dotnet/Simple.Sauce/SauceOptions.cs

```diff
@@ -55,83 +59,114 @@ public SauceOptions()
-                                                edgeVersion);
-            ConfiguredEdgeOptions.BrowserVersion = edgeVersion.Value;
-            ConfiguredEdgeOptions.PlatformName = DEFAULT_PLATFORM;
+            var sauceUsername = Environment.GetEnvironmentVariable("SAUCE_USERNAME");
+            var sauceAccessKey = Environment.GetEnvironmentVariable("SAUCE_ACCESS_KEY");
+
+            return new Dictionary<string, object>
+            {
+                ["username"] = sauceUsername,
+                ["accessKey"] = sauceAccessKey
+            };
+        }
+
+        private Dictionary<string, object> AddOptions()
+        {
+            var options = JsonConvert.SerializeObject(this, JsonUtils.SerializerSettings());
+            var sauceOptions = JsonConvert.DeserializeObject<Dictionary<string, object>>(options, JsonUtils.SerializerSettings());
+
+            foreach (var timeout in Timeout?.ToDictionary())
+                sauceOptions.Add(timeout.Key, timeout.Value);
+
+            if (CustomData != null)
+                sauceOptions.Add("custom-data", CustomData);
+
+            return sauceOptions;
         }
 
-        public void WithChrome()
+        private void ConfigureOptions(string version = DEFAULT_BROWSER_VERSION, string platform = DEFAULT_PLATFORM)
         {
-            ConfiguredChromeOptions.BrowserVersion = DEFAULT_BROWSER_VERSION;
-            ConfiguredChromeOptions.PlatformName = DEFAULT_PLATFORM;
+            ConfiguredOptions.BrowserVersion = version;
+            ConfiguredOptions.PlatformName = platform;
         }
 
-        public void WithChrome(string chromeVersion)
+        public SauceOptions WithChrome(string version = DEFAULT_BROWSER_VERSION, string platform = DEFAULT_PLATFORM)
         {
-            ConfiguredChromeOptions.BrowserVersion = chromeVersion;
+            ConfiguredOptions = new ChromeOptions();
+            ConfigureOptions(version, platform);
+            return this;
+        }
+
+        public SauceOptions WithEdge() => WithEdge(EdgeVersion.Latest);
+        public SauceOptions WithEdge(EdgeVersion edgeVersion)
+        {
+            if (edgeVersion is null)
+                throw new ArgumentNullException("Please supply a valid Edge Version. You supplied an invalid value => " +
+                                                edgeVersion);
+
+            ConfiguredOptions = new EdgeOptions();
+            ConfigureOptions(edgeVersion.Value);
+            return this;
         }
 
-        public void WithSafari()
+        public SauceOptions WithFirefox(string version = DEFAULT_BROWSER_VERSION, string platform = DEFAULT_PLATFORM)
         {
-            WithSafari(DEFAULT_BROWSER_VERSION);
+            ConfiguredOptions = new FirefoxOptions();
+            ConfigureOptions(version, platform);
+            return this;
         }
 
-        public void WithSafari(string safariVersion)
+        public SauceOptions WithSafari(string version = DEFAULT_BROWSER_VERSION)
```

Comment:
Codacy found an issue: [Use the overloading mechanism instead of the optional parameters.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6319844)